services/121-service/src/payments/fsp-integration/safaricom/safaricom.api.service.ts

```diff
@@ -0,0 +1,79 @@
+import { Injectable } from '@nestjs/common';
+import { TokenSet } from 'openid-client';
+import { CustomHttpService } from '../../../shared/services/custom-http.service';
+import {
+  SafaricomAuthResponseDto,
+  SafaricomTransferResponseDto,
+} from './dto/safaricom-load-response.dto';
+
+@Injectable()
+export class SafaricomApiService {
+  public tokenSet: TokenSet;
+
+  public constructor(private readonly httpService: CustomHttpService) {}
+
+  public async authenticate(): Promise<string> {
+    const consumerKey = process.env.SAFARICOM_CONSUMER_KEY;
+    const consumerSecret = process.env.SAFARICOM_CONSUMER_SECRET;
+    const accessTokenUrl = `${process.env.SAFARICOM_API_URL}/oauth/v1/generate?grant_type=client_credentials`;
+    const auth = Buffer.from(`${consumerKey}:${consumerSecret}`).toString(
+      'base64',
+    );
+
+    // Check expires_at
+    if (this.tokenSet && this.tokenSet.expires_at > Date.now()) {
+      // Return cached token
+      return this.tokenSet.access_token;
+    } else {
+      // If not valid, request new token
+      try {
+        const headers = [{ name: 'Authorization', value: `Basic ${auth}` }];
+
+        const { data } = await this.httpService.get<SafaricomAuthResponseDto>(
+          `${accessTokenUrl}`,
+          headers,
+        );
+
+        const datetime = new Date();
+        // Cache tokenSet and expires_at
+        const tokenSet = new TokenSet({
+          access_token: data.access_token,
+          expires_at: datetime.setMinutes(datetime.getMinutes() + 55),
+        });
+
+        this.tokenSet = tokenSet;
+
+        return tokenSet.access_token;
+      } catch (error) {
+        console.log(error, 'authenticate');
+        console.error('Failed to make OAuth Access Token payment API call');
+        return this.tokenSet.access_token;
```

Comment:
I'm doubting if this should return an access_token here. The token is apparently is not valid so I'm not sure why we would return that here.

Reply:
You are right and now I will remove it.